File: src/main/java/com/openlattice/chronicle/configuration/ChronicleConfiguration.java
package com.openlattice.chronicle.configuration;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.kryptnostic.rhizome.configuration.annotation.ReloadableConfiguration;
import java.util.Objects;

@ReloadableConfiguration(uri="chronicle.yaml")
public class ChronicleConfiguration {
    private final String USER_KEY = "user";
    private final String PASSWORD_KEY = "password";

    private final String user;
    private final String password;

    public ChronicleConfiguration(
            @JsonProperty( USER_KEY ) String user,
            @JsonProperty( PASSWORD_KEY ) String password ) {
        this.user = user;
        this.password = password;
    }

    @JsonProperty( USER_KEY )
    public String getUser() {
        return user;
    }

    @JsonProperty( PASSWORD_KEY )
    public String getPassword() {
        return password;
    }

    @Override public boolean equals( Object o ) {
        if ( this == o ) { return true; }
        if ( !( o instanceof ChronicleConfiguration ) ) { return false; }
        ChronicleConfiguration that = (ChronicleConfiguration) o;
        return Objects.equals( user, that.user ) &&
                Objects.equals( password, that.password );
    }

    @Override public int hashCode() {
        return Objects.hash( user, password );
    }

    @Override public String toString() {
        return "ChronicleConfiguration{" +
                "user='" + user + '\'' +
                ", password='" + password + '\'' +
                '}';
    }
}

File: src/main/kotlin/com/openlattice/chronicle/configuration/ChronicleConfiguration.kt
package com.openlattice.chronicle.configuration

import com.fasterxml.jackson.annotation.JsonProperty
import com.geekbeast.rhizome.configuration.Configuration
import com.geekbeast.rhizome.configuration.ConfigurationKey
import com.geekbeast.rhizome.configuration.SimpleConfigurationKey
import com.geekbeast.rhizome.configuration.configuration.annotation.ReloadableConfiguration

const val BUCKET_NAME = "bucketName"
const val REGION_NAME = "regionName"
const val TIME_TO_LIVE = "timeToLive"
const val ACCESS_KEY_ID = "accessKeyId"
const val SECRET_ACCESS_KEY = "secretAccessKey"
const val STORAGE_CONFIGURATION = "storageConfiguration"

@ReloadableConfiguration(uri = "chronicle.yaml")
data class ChronicleConfiguration(
    @JsonProperty(BUCKET_NAME) val bucketName: String,
    @JsonProperty(REGION_NAME) val regionName: String,
    @JsonProperty(TIME_TO_LIVE) val timeToLive: Long,
    @JsonProperty(ACCESS_KEY_ID) val accessKeyId: String,
    @JsonProperty(SECRET_ACCESS_KEY) val secretAccessKey: String,
    @JsonProperty(STORAGE_CONFIGURATION) val storageConfiguration: ChronicleStorageConfiguration = ChronicleStorageConfiguration()
) : Configuration {

    companion object {
        @JvmField
        val key = SimpleConfigurationKey("chronicle.yaml")
    }

    override fun getKey(): ConfigurationKey {
        return ChronicleConfiguration.key
    }
}
